### PR TITLE
[33677] Reduce complexity of boolean filters

### PR DESCRIPTION
**Pull request by bsatarnejad**
adding toggle switch for for boolean filter to make it simpler

https://community.openproject.com/projects/openproject/work_packages/33677/activity

### CODE REVIEW

File: frontend/src/app/components/filters/filter-boolean-value/filter-boolean-value.component.ts [hunk 1]
```diff
@@ -33,7 +33,8 @@ import {QueryFilterInstanceResource} from 'core-app/modules/hal/resources/query-
```
Comment (openprojectci) [inline frontend/src/app/components/filters/filter-boolean-value/filter-boolean-value.component.ts:33]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/a6d121d5804683a31696c23069b691a8cfc94e9d/frontend/src/app/components/filters/filter-boolean-value/filter-boolean-value.component.ts#L33"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

File: frontend/src/app/components/filters/query-filter/query-filter.component.html [hunk 2]
```diff
@@ -68,7 +75,7 @@
                            [shouldFocus]="shouldFocus"
                            [filter]="filter"></filter-string-value>
 
-      <filter-boolean-value *ngSwitchCase="'[1]Boolean'"
+      <filter-boolean-value *ngSwitchCase="'[1]Boolean'"  hidden
```
Comment (oliverguenther) [inline frontend/src/app/components/filters/query-filter/query-filter.component.html]: Why did you add a `hidden` here instead of removing this value alltogether?

File: frontend/src/app/components/filters/query-filter/query-filter.component.html [hunk 3]
```diff
@@ -16,6 +16,7 @@
     </label>
 
     <select required
+            *ngIf="valueType !== '[1]Boolean'"
```
Comment (oliverguenther) [inline frontend/src/app/components/filters/query-filter/query-filter.component.html:19]: Please turn this check into a method on the `QueryFilterComponent` something like `operatorVisible(valueType)`, this will make it a bit more clear when the operator is hidden.

**Review by oliverguenther — CHANGES_REQUESTED** [review]
Hi @bsatarnejad , the changes look great overall. I like that you added a toggle to it.

We need to put this toggle into its own component so we can reuse it, as I have many more places in the application where I'd like to use it.

For that, please:

- Move it into its own component folder under `frontend/src/app/modules/common/`. Angular Material has such a component and calls it `slide-toggle`. that could be a fitting name. What do you think?
- Encapsulate the styles and template in that folder
- Add input/outputs for the value changing
  - An input for the ID attribute
  - An input for the set value for `[ngModel]`
  - An output for `(ngModelChange)`
- Add a spec to the component that tests the input/outputs


I also have a tiny bit of styling remarks: I think the height of the toggle is a bit too large. What do you think?

File: frontend/src/app/components/filters/query-filter/query-filter.component.html (hunk 2 above)
Comment (bsatarnejad) [reply frontend/src/app/components/filters/query-filter/query-filter.component.html]: Hi @oliverguenther , thanks for your feedback, i agree with you to create a separate component to be used for the whole project. we can reduce the size of it to be more user-friendly.

When i removed filter-boolean-value component, i have an error in  filter-toggled-multiselect-value.component.ts file where there is function    

 private fetchAllowedValues() {
    if ((this.filter.currentSchema!.values!.allowedValues as CollectionResource)['$load']) {
      this.loadAllowedValues();
    } else {
      this.availableOptions = (this.filter.currentSchema!.values!.allowedValues as HalResource[]);
    }
  }
,   (this.filter.currentSchema!.values!.allowedValues as CollectionResource) is null at the time of selecting a boolean filter. 

filter-toggled-multiselect-value component is located after filter-boolean-value component in query-filter.component.html .

I just set hidden to it to ignore this error at the moment.

File: frontend/src/app/components/filters/query-filter/query-filter.component.html (hunk 3 above)
Comment (bsatarnejad) [reply frontend/src/app/components/filters/query-filter/query-filter.component.html:19]: yes that's a great idea. i will go through this.

**Comment by openprojectci** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/57f6bf3cb4fdf8cd9d203d1a9f6e280983d7e137/frontend/src/app/components/filters/filter-boolean-value/filter-boolean-value.component.ts#L33">frontend/src/app/components/filters/filter-boolean-value/filter-boolean-value.component.ts#L33</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/57f6bf3cb4fdf8cd9d203d1a9f6e280983d7e137/frontend/src/app/components/filters/query-filter/query-filter.component.ts#L37">frontend/src/app/components/filters/query-filter/query-filter.component.ts#L37</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by oliverguenther — APPROVED** [review]
Great stuff, I really like the functionality :+1:
